### PR TITLE
Drafting load test for Pubsub to Bigquery

### PR DESCRIPTION
**Pull request by pabloem**

### CODE REVIEW

**Comment by pabloem** [comment]
r: @bvolpato 

I would say that the Java PR / Build and Unit Test breakages are from a bug in the build script that will not include internal subproject depenfdencies if they're not changed (e.g. should include `metadata` and `integration-testing` even if they're not changed)

**Comment by pabloem** [comment]
r: @pranavbhandari24 please TAL : )

**Review by pranavbhandari24 — COMMENTED** [review]
I reviewed the Load test related files. You might want someone else to review Syndeo related things.

Also, as we had discussed you might want to use Streaming Data Generator template for generating fake data. It supports generating fake data to a Pub/Sub topic in multiple formats (which seems like what you want). Look at `PubsubToBigqueryLT.java` for an example.

File: it/src/main/java/com/google/cloud/teleport/it/pubsub/DefaultPubsubResourceManager.java [hunk 1]
```diff
@@ -260,6 +263,15 @@ public Builder credentialsProvider(CredentialsProvider credentialsProvider) {
     }
 
     public DefaultPubsubResourceManager build() throws IOException {
+      if (credentialsProvider == null) {
```
[the remark this answers is not in the document]
Comment (pabloem) [reply it/src/main/java/com/google/cloud/teleport/it/pubsub/DefaultPubsubResourceManager.java:266]: this allows `DefaultPubsubResourceManager` to automatically infer a credentials provider (much like other Resource managers)

File: syndeo-template/src/test/java/com/google/cloud/syndeo/perf/SyndeoPubsubToBigQueryLT.java [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+  @Before
+  public void beforeTest() throws IOException {
+    bigquery = DefaultBigQueryResourceManager.builder("syndeo-pubsub-bq-lt", PROJECT).build();
+    pubsub = DefaultPubsubResourceManager.builder("syndeo-pubsub-bq-lt", PROJECT).build();
+    testStart = Instant.now();
+  }
+
+  @After
+  public void afterTest() {
+    if (pubsub != null) pubsub.cleanupAll();
```
[the remark this answers is not in the document]
Comment (pabloem) [reply syndeo-template/src/test/java/com/google/cloud/syndeo/perf/SyndeoPubsubToBigQueryLT.java]: nice. Thanks!

File: syndeo-template/src/test/java/com/google/cloud/syndeo/perf/SyndeoPubsubToBigQueryLT.java [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+      syndeoResult.cancel();
+      fail("Data generator job failed to start. Unable to perform test.");
+    }
+
+    while (true) {
+      Long inputElements =
+          Math.min(getElementsProcessed(generatorResult.metrics()), testConfig.getNumRows());
+      Long syndeoProcessed = getElementsProcessed(syndeoResult.metrics());
+      int testRuntime = new Period(testStart, Instant.now()).toStandardMinutes().getMinutes();
+      System.out.printf(
```
[the remark this answers is not in the document]
Comment (pabloem) [reply syndeo-template/src/test/java/com/google/cloud/syndeo/perf/SyndeoPubsubToBigQueryLT.java]: done

**Comment by pabloem** [comment]
>Also, as we had discussed you might want to use Streaming Data Generator template for generating fake data. It supports generating fake data to a Pub/Sub topic in multiple formats (which seems like what you want). Look at PubsubToBigqueryLT.java for an example.

I think for now I'd like to stick to this method. I can make the change as a follow up if necessary. Is that okay?

**Comment by pabloem** [comment]
@pranavbhandari24 addressed all comments : ))))))

File: it/src/main/java/com/google/cloud/teleport/it/pubsub/DefaultPubsubResourceManager.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (pranavbhandari24) [reply it/src/main/java/com/google/cloud/teleport/it/pubsub/DefaultPubsubResourceManager.java:266]: I'm not sure if the RM should be responsible for inferring credentials. I think currently this is done in the test itself and the credentials are passed to the RM using the builder methods.

@bvolpato any thoughts?

**Review by xianhualiu — COMMENTED** [review]
following two new files look good. to me:

- syndeo-template/src/test/java/com/google/cloud/syndeo/perf/SyndeoPubsubToBigQueryLT.java
- syndeo-template/src/main/java/com/google/cloud/syndeo/transforms/pubsub/SyndeoPubsubReadSchemaTransformProvider.java

File: it/src/main/java/com/google/cloud/teleport/it/pubsub/DefaultPubsubResourceManager.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (bvolpato) [reply it/src/main/java/com/google/cloud/teleport/it/pubsub/DefaultPubsubResourceManager.java:266]: Inferring will make it easier to use, but there's also a risk of using the wrong project unintentionally when using the ResourceManager if we don't make it explicit.

I guess if we think about more generic ResourceManagers when it goes to Beam, not making assumption on the environment would be better? 

For other templates we are declaring in a base class (this is based on `LoadTestBase`):

```
  protected static final Credentials CREDENTIALS = TestProperties.googleCredentials();
  protected static final CredentialsProvider CREDENTIALS_PROVIDER =
      FixedCredentialsProvider.create(CREDENTIALS);
```

And then just use that in RM's builder. 

Maybe that's a good balance here? TBH I have no strong opinion as I see pros and cons

[the remark this answers is not in the document]
Comment (pabloem) [reply it/src/main/java/com/google/cloud/teleport/it/pubsub/DefaultPubsubResourceManager.java:266]: I'm happy to go with whatever other RMs are doing. I'll try and follow that.

[the remark this answers is not in the document]
Comment (pabloem) [reply it/src/main/java/com/google/cloud/teleport/it/pubsub/DefaultPubsubResourceManager.java:263]: done.

**Comment by pabloem** [comment]
I'm adding the lgtm label to start moving the PR forward, but happy to address more comments if there are.

**Comment by pabloem** [comment]
@pranavbhandari24 let me know if there's any other comments : )

**Comment by pranavbhandari24** [comment]
LGTM, Thanks! 

Please squash commits

**Comment by pabloem** [comment]
@pranavbhandari24 done! can you please approve the merge:? : )

**Review by pranavbhandari24 — APPROVED** [review]

**Pull request merged**